### ISSUE TITLE
docs: clarification on DMatrix CSR return

### ISSUE THREAD

**Issue by ExpandingMan** [dmlc/xgboost#8459]
I'm in the process of implementing the new function `XGDMatrixGetDataAsCSR` in the XGBoost.jl Julia wrapper.

I'm a bit confused about what this is supposed to be returning in the presence of null values.  Values passed to the `DMatrix` as `NaN32` do not show up in the sparse matrix, but values passed as `0.0f0` *are* there.  This raises the question of what is the default value of the resulting sparse matrix: is it null or zero?

Both options seem to present problems:
- If the former, there is a major bug in the Python implementation which returns a `scipy.sparse.csr_matrix` which has default values of `0`.
- If the latter, either the `DMatrix` constructors or the matrix returned by `XGDMatrixGetDataAsCSR` is faulty because `NaN` values are getting converted to `0`'s somewhere.

Can someone provide clarification on the intended behavior?  Thanks!

**Comment by trivialfis** [comment]
Hi, DMatrix constructor has a parameter called `missing`, which is defaulted to `NaN`. Any entries with `NaN` value will be removed during the construction of DMatrix.

> This raises the question of what is the default value of the resulting sparse matrix: is it null or zero?

It's the argument passed to `missing` .

> If the former, there is a major bug in the Python implementation which returns a scipy.sparse.csr_matrix which has default values of 0.

``` python
from scipy import sparse
import xgboost as xgb
from numpy import testing
import numpy as np

values = [0.0, 1.0, 0.0]
indptr = [0, 3]
indices = [0, 1, 2]
csr0 = sparse.csr_matrix((values, indices, indptr), shape=(1, 3))

X = xgb.DMatrix(csr0)
csr1 = X.get_data()
testing.assert_allclose(np.array(values, dtype=np.float32), csr1.data)

```

**Comment by ExpandingMan** [comment]
I'm a bit confused.  The default value for `sparse.csr_matrix` is always `0`, is it not?  If that's the case, the returned `csr_matrix` will not match the input data unless the default value passed on construction is `0`.  Indeed it looks to me like in the code you provided you are only comparing the non-zero value data and not checking the elements of the matrix which are not given explicitly as part of `values`.

Also, is there a way to retrieve the default value from the `DMatrix`?  I don't see a function that does that in the API.

**Comment by trivialfis** [comment]
Let's distinguish "zero" and "missing". zero means the real number 0 on the number line, while `missing` means the omitted values in a sparse matrix, which may or may not be 0. Another way of saying missing is "unknown". One can represent the concept of "unknown" using any value, `NaN`, `None`, `Inf`, `0`, `-1`, or whatever the practitioner sees fit.

For instance, given a dataset that contains a predictor storing the temperature of different locations, 0 is a valid temperature value. For locations where we haven't been able to collect the temperature, we might express it as `NaN` or `Inf` in a table:

``` python
# We don't know the temperature of location 4, use NaN to represent this piece of missing information.
X = pd.Dataframe(
    {
        "temperature": [0, 20, 13, 4, np.nan],
        "location_code": [0, 1, 2, 3, 4],
        "has_dog": [True, False, False, True, True],
    }
)

m = DMatrix(X, missing=np.nan)
```

>  The default value for sparse.csr_matrix is always 0, is it not

It can be something else like `NaN` as in the example in the previous comment, 0 is a valid value for the `csr_matrix` and `DMatrix`.

> Indeed it looks to me like in the code you provided you are only comparing the non-zero value data and 

I was comparing 0 values as well, you can print out the content of CSR matrix if you wish. But indeed I'm not comparing the "unknown" value directly, anything that's not in the input csr_matrix is unknown.

> not checking the elements of the matrix which are not given explicitly as part of values.

Indeed. But I'm not sure what's there to compare. The known values are correctly returned, and the rest of the values are unknown.

> Also, is there a way to retrieve the default value from the DMatrix? 

No.

In short, just take `DMatrix` as a generalized CSR matrix that a user can specify what value to remove.

``` python
# A dense matrix with shape `(1, 6)` where `-1` means that the entry is invalid/unknown/missing.
array = np.array([1, 2, 3, 4, 5, -1])
# m is a CSR matrix that stores values `[1, 2, 3, 4, 5]` and shape `(1, 6)` with the value -1 being omitted.
m = DMatrix(array, missing=-1)
```

**Comment by trivialfis** [comment]
> indexing by values not listed in v, for example A[1,0] == 0

Indeed, this is inconsistent with the semantics for DMatrix. If `A[1, 0]` is missing, then indexing into `A[1, 0]` should throw an error instead of returning 0 if the semantic is respected.

**Comment by trivialfis** [comment]
Closing now. We can revisit if there's a better alternative to scipy sparse.
